File: benchmarks/crc32/ref.js
// crc32 reference, Node. Bitwise, no table, no library.
const fs = require("fs");
const buf = fs.readFileSync(process.argv[2]);
let crc = 0xFFFFFFFF;
for (let i = 0; i < buf.length; i++) {
  let c = (crc ^ buf[i]) >>> 0;
  for (let k = 0; k < 8; k++) c = (c & 1) === 1 ? ((c >>> 1) ^ 0xEDB88320) >>> 0 : c >>> 1;
  crc = c;
}
crc = (crc ^ 0xFFFFFFFF) >>> 0;
console.log("bytes " + buf.length);
console.log("crc32 " + crc);

File: benchmarks/crc32/ref.rb
# crc32 reference, Ruby. Bitwise, no table, not Zlib.crc32.
buf = File.binread(ARGV[0])
crc = 0xFFFFFFFF
buf.each_byte do |b|
  c = crc ^ b
  8.times { c = (c & 1) == 1 ? ((c >> 1) ^ 0xEDB88320) : (c >> 1) }
  crc = c
end
crc ^= 0xFFFFFFFF
puts "bytes #{buf.bytesize}"
puts "crc32 #{crc}"
